### PR TITLE
[libsolv] Add new port

### PR DESCRIPTION
**Pull request by pavelzw**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [X] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [X] The license declaration in `vcpkg.json` matches what upstream says.
- [X] The installed as the "copyright" file matches what upstream says.
- [X] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by jimwang118** [comment]
Note: I will be converting your PR to draft status. The above suggested changes are only recommendations. If you are willing to adopt them, please click "ready for review" after making the modifications. If you do not wish to make any changes, please click "ready for review" directly. That way, I can be aware that you've responded since you can't modify the tags.

File: ports/libsolv/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DDISABLE_SHARED=${DISABLE_SHARED}
+        -DENABLE_STATIC=${BUILD_STATIC_LIBS}
+        -DMULTI_SEMANTICS=ON
+        ..
```
Comment (jimwang118) [inline ports/libsolv/portfile.cmake:49]:
```suggestion
```

**Review by jimwang118 — CHANGES_REQUESTED** [review]

**Comment by jimwang118** [comment]
The following error occurs when compiling locally:
[config-x64-windows-out.log](https://github.com/microsoft/vcpkg/files/11411620/config-x64-windows-out.log)

**Comment by pavelzw** [comment]
The bindings are not working for windows due to [only-posix-headers](https://github.com/openSUSE/libsolv/blob/5dc4f81fda718df26f8f5dda343a71d34cf555b4/bindings/solv.i#L712) so I deactivated them.

File: ports/libsolv/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+if (ENABLE_RUBY)
+    vcpkg_find_acquire_program(RUBY)
+    list(APPEND FEATURE_OPTIONS "-DRUBY_EXECUTABLE=${RUBY}")
+endif()
+
+if(WIN32)
+    list(APPEND FEATURE_OPTIONS "-DWITHOUT_COOKIEOPEN=ON")
+endif()
+
+vcpkg_configure_cmake(
```
Comment (jimwang118) [inline ports/libsolv/portfile.cmake:59]:
```suggestion
vcpkg_cmake_configure(
```

File: ports/libsolv/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DDISABLE_SHARED=${DISABLE_SHARED}
+        -DENABLE_STATIC=${BUILD_STATIC_LIBS}
+        -DMULTI_SEMANTICS=ON
+)
+
+vcpkg_install_cmake()
```
Comment (jimwang118) [inline ports/libsolv/portfile.cmake:69]:
```suggestion
vcpkg_cmake_install()
```

**Review by jimwang118 — CHANGES_REQUESTED** [review]

**Comment by jimwang118** [comment]
Use the compilation command: `.\vcpkg.exe install libsolv[*]:x64-windows --allow-unsupported`, the following error occurs:
```
C:\PROGRA~1\MICROS~4\2022\ENTERP~1\VC\Tools\MSVC\1434~1.319\bin\Hostx64\x64\cl.exe -DENABLE_BZIP2_COMPRESSION=1 -DENABLE_COMPS=1 -DENABLE_CONDA=1 -DENABLE_DEBIAN =1 -DENABLE_HELIXREPO=1 -DENABLE_LZMA_COMPRESSION=1 -DENABLE_ZLIB_COMPRESSION=1 -DENABLE_ZSTD_COMPRESSION=1 -DMULTI_SEMANTICS=1 -DWITHOUT_COOKIEOPEN=1 -D_CRT_SECURE_NO_WARNINGS -D_FILE_OFFSET_BITS=64 -Dbindings_python_EXPORTS -IF:\libsolv\installed\x64-windows\include -IF: \libsolv\buildtrees\libsolv\src\0.7.24-f99b2a1376.clean -IF:\libsolv\buildtrees\libsolv\src\0.7.24-f99b2a1376.clean\src -IF:\libsolv\buildtrees\libsolv\src\0.7 .24-f99b2a1376.clean\ext -IF:\libsolv\buildtrees\libsolv\x64-windows-dbg -IF:\libsolv\buildtrees\libsolv\x64-windows-dbg\src -IF:\libsolv\installed\x64- windows\include\python3.10 /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /MP /wd4244 /wd4267 -fPIC /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1 -Wno-unused /showIncludes /Fobindings\python\ CMakeFiles\bindings_python.dir\solv_python.c.obj /Fdbindings\python\CMakeFiles\bindings_python.dir\ /FS -c F:\libsolv\buildtrees\libsolv\x64-windows-dbg\bindings\python\solv_python.c
cl : Command line error D8021 : invalid numeric argument '/Wno-unused'
```

**Comment by dg0yt** [comment]
` --allow-unsupported`  means unsupported. Why should this block a PR?

File: ports/libsolv/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+if(WIN32)
+    list(APPEND FEATURE_OPTIONS "-DWITHOUT_COOKIEOPEN=ON")
+endif()
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DDISABLE_SHARED=${DISABLE_SHARED}
```
Comment (dg0yt) [inline ports/libsolv/portfile.cmake:64]: Can't you use `BUILD_STATIC_LIBS` here?

**Comment by pavelzw** [comment]
`/Wno-unused` is a GCC compiler flag that ist added in the bindings [CMakeLists](https://github.com/openSUSE/libsolv/blob/5dc4f81fda718df26f8f5dda343a71d34cf555b4/bindings/python/CMakeLists.txt#L37). Since the bindings are meant to only be compiled on GCC and Linux (also mentioned [here](https://github.com/microsoft/vcpkg/pull/31275#issuecomment-1537163981)), I suggest we leave it unsupported. This would be part of a slightly bigger upstream change…

File: ports/libsolv/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    vcpkg_find_acquire_program(RUBY)
+    list(APPEND FEATURE_OPTIONS "-DRUBY_EXECUTABLE=${RUBY}")
+endif()
+
+if(WIN32)
+    list(APPEND FEATURE_OPTIONS "-DWITHOUT_COOKIEOPEN=ON")
+endif()
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
```
Comment (dg0yt) [inline ports/libsolv/portfile.cmake:60]: Double quotes for the argument.

File: ports/libsolv/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        python ENABLE_PYTHON
+        ruby ENABLE_RUBY
+        comps ENABLE_COMPS
+        helixrepo ENABLE_HELIXREPO
+        debian ENABLE_DEBIAN
+        cudfrepo ENABLE_CUDFREPO
+        conda ENABLE_CONDA
+        lzma-compression ENABLE_LZMA_COMPRESSION
+        bzip2-compression ENABLE_BZIP2_COMPRESSION
+        zstd-compression ENABLE_ZSTD_COMPRESSION
```
Comment (dg0yt) [inline ports/libsolv/portfile.cmake:25]: Most ports wouldn't use the `-compression` suffix for these features but I don' t ask for changes.

**Review by dg0yt — DISMISSED** [review]

File: ports/libsolv/portfile.cmake (hunk 6 above)
Comment (pavelzw) [reply ports/libsolv/portfile.cmake:25]: Alright then let's get rid of it; I don't have a strong opinion on this.

**Review by jimwang118 — DISMISSED** [review]

**Comment by jimwang118** [comment]
After compiling, how can I call this library? It is recommended to provide a usage instruction, so that users will be more convenient to use.

**Comment by dg0yt** [comment]
> After compiling, how can I call this library? It is recommended to provide a usage instruction, so that users will be more convenient to use.

This could be documented for pkgconfig, exampe: port mujs.

**Comment by pavelzw** [comment]
How about this usage file?

```
The package libsolv provides CMake targets:

    find_package(libsolv REQUIRED)
    include_directories(${LibSolv_INCLUDE_DIRS})
    target_link_libraries(main PRIVATE ${LibSolv_LIBRARIES})
```

**Comment by dg0yt** [comment]
> How about this usage file?
> 
> ```
> The package libsolv provides CMake targets:
> 
>     find_package(libsolv REQUIRED)
>     include_directories(${LibSolv_INCLUDE_DIRS})
>     target_link_libraries(main PRIVATE ${LibSolv_LIBRARIES})
> ```

AFAICS there is no implementation for `find_package(libsolv REQUIRED)`, so it doesn't work.
That's why I deliberately pointed at another solution.

**Comment by pavelzw** [comment]
image

When I installed it locally, I found the `FindLibSolv.cmake`, `find_package(LibSolv REQUIRED)` should work, right?

**Comment by dg0yt** [comment]
So a find module exists, and the spelling matches now. But AFAICT, this find module isn't in the search path, so it isn't found. You need to add a wrapper or config file which delegates to the find module in order to make it work.

**Comment by jimwang118** [comment]
Note: I will be converting your PR to draft status. The above suggested changes are only recommendations. If you are willing to adopt them, please click "ready for review" after making the modifications. If you do not wish to make any changes, please click "ready for review" directly. That way, I can be aware that you've responded since you can't modify the tags.

**Comment by jimwang118** [comment]
@pavelzw Thanks for your PR, Is work still being done for this PR?

**Comment by jimwang118** [comment]
Closing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.

**Pull request closed without merging**